README
CIHM-METS-App

Application used to generate specific metadata components used during the generation of METS records for Canadiana.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

Documentation will come later.  For now, please see:
  https://github.com/c7a/CIHM-METS-App

LICENSE AND COPYRIGHT

Copyright (C) 2017 Russell McOrmond

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for CIHM-METS-app

0.01    2017-05-29 15:20:00
        Moving code from private subversion repository, and wrapping in environment created by module-starter



Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'CIHM::METS::App',
    AUTHOR           => q{Russell McOrmond <russell.mcormond@canadiana.ca>},
    VERSION_FROM     => 'lib/CIHM/METS/App.pm',
    ABSTRACT_FROM    => 'lib/CIHM/METS/App.pm',
    LICENSE          => 'Artistic_2_0',
    PL_FILES         => {},
    "EXE_FILES" => [
      "bin/mets"
    ],
    MIN_PERL_VERSION => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        'MooseX::App'              => 0,
	'Try::Tiny' => 0,
	'Data::Dumper' => 0,
	'Module::Load::Conditional' => 0,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'CIHM-METS-App-*' },
);


lib/CIHM/METS/App.pm
package CIHM::METS::App;

use MooseX::App;
use Data::Dumper;
use Try::Tiny;
use Module::Load::Conditional qw[can_load check_install requires];

=head1 NAME

CIHM::METS::App - Command-line tool for generating/manipulating components of METS records.

=head1 VERSION

Version 0.02

=cut

our $VERSION = '0.02';

=head1 AUTHOR

Sascha Adler, C<< <sascha.adler at canadiana.ca> >>
Russell McOrmond, C<< <russell.mcormond at canadiana.ca> >>
Julienne Pascoe, C<< <julienne.pascoe at canadiana.ca> >>

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/c7a/CIHM-METS-App>.  We will be notified, and then you'll
automatically be notified of progress on your bug as we make changes.

=cut

option 'dump' => (
  is => 'rw',
  isa => 'Bool',
  documentation => q[Dump information to console rather than send to CouchDB],
);

has 'WIP' => (
    is => 'rw',
    lazy => 1,
    builder => '_build_WIP',
);

sub _build_WIP {
    my $self = shift;

    my $WIP;
    if (!$self->dump && can_load(modules => {
        'CIHM::WIP' => undef
                 })) {
        $WIP = CIHM::WIP->new($self->conf);
    }
    if (!$WIP) {
        warn "Not using CIHM::WIP\n";
    }
    return $WIP;
};

has 'objid' => (
    is => 'rw',
    isa => 'Str'
);

has 'myconfig' => (
    is => 'rw',
    isa => 'HashRef'
);

has 'identifier' => (
    is => 'rw',
    isa => 'Str'
);

option 'conf' => (
  is => 'rw',
  isa => 'Str',
  default => "/etc/canadiana/wip/wip.conf",
  documentation => q[An option that specifies where you can find a config file if not default],
);

parameter 'configid' => (
  is => 'rw',
  isa => 'Str',
  required => 1,
  documentation => q[The configuration ID (Example: heritage)],
);

option 'depositor' => (
  is => 'rw',
  isa => 'Str',
  documentation => q[The depositor (Example: oocihm)],
);

sub set_objid {
    my ($self,$identifier) = @_;

    #trim whitespace
    $identifier =~ s/^\s+|\s+$//g;

    # If we have CIHM::WIP
    if ($self->WIP) {
        my $objid=$self->WIP->i2objid($identifier,$self->configid);
        if (!$self->WIP->objid_valid($objid)) {
            die "$objid not valid OBJID\n";
        }
        $self->objid($objid);
    } else {
        # Set identifier as objectID as we can't use translation without WIP
        $self->objid($identifier);
    }
}


sub setup {
    my $self = shift;

    # If we have CIHM::WIP
    if ($self->WIP) {
        if (!$self->depositor) {
            my $configdocs=$self->WIP->configdocs ||
                die "Can't retrieve configuration documents\n";

            my $myconfig=$configdocs->{$self->configid} ||
                die $self->configid." is not a valid configuration id\n";

            $self->myconfig($myconfig);

            my $depositor=$myconfig->{depositor} ||
                die "Depositor not set for ".$self->configid."\n";

            $self->depositor($depositor);
        }
    }
    if ($self->identifier) {
        $self->set_objid($self->identifier);
    }
    if (!$self->depositor) {
        die "--depositor not set (and not available from configid)\n";
    }
    
}

sub itemlabel {
    my $self = shift;

    if ($self->myconfig) {
        return $self->myconfig->{itemlabel};
    }
}


sub dataDump {
    my ($self,$uid,$params) = @_;

    print Data::Dumper->Dump([$params], [$uid]);
}

sub couchSend {
    my ($self,$params) = @_;

    my $uid=join('.',$self->depositor,$self->objid);


    if (!$self->WIP) {
        return $self->dataDump($uid,$params);
    }

    my ($res, $revision);

    my $wipmeta=$self->WIP->wipmeta;
    my $updatedoc={};
    if ($params->{label}) {
        $updatedoc->{label}=$params->{label};
    }

#    print "couchSend: " . Dumper($params)."\n";

    # This encoding makes variables available as form data
    $wipmeta->type("application/x-www-form-urlencoded");
    $res = $wipmeta->post("/".$wipmeta->{database}."/_design/tdr/_update/basic/".$uid, $updatedoc, {deserializer => 'application/json'});

    if ($res->code != 201 && $res->code != 200) {
        die "_update/basic/$uid POST return code: " . $res->code . "\n";
    }
    $revision=$res->response->header('X-Couch-Update-Newrev');


    if ($params->{clabels} || $params->{dmdsec} || $params->{metadata}) {
        if (!$revision) {
            $res = $wipmeta->head("/".$wipmeta->database."/$uid",{}, {deserializer => 'application/json'});
            if ($res->code == 200) {
                $revision=$res->response->header("etag");
                $revision =~ s/^\"|\"$//g
            } else {
                die "couchSend($uid) HEAD return code: ".$res->code."\n"; 
            }
        }
        if ($params->{clabels}) {
            $revision=$self->put_attachment($uid,$revision,'labels.json','application/json',$params->{clabels});
        }
        if ($params->{dmdsec}) {
            $revision=$self->put_attachment($uid,$revision,'dmd.xml','application/xml',$params->{dmdsec});
        }
        if ($params->{metadata}) {
            $revision=$self->put_attachment($uid,$revision,'metadata.xml','application/xml',$params->{metadata});
        }
    }
}

sub put_attachment {
    my ($self,$uid,$revision,$name,$type,$data) = @_;
    my $wipmeta=$self->WIP->wipmeta;

    $wipmeta->clear_headers;
    $wipmeta->set_header('If-Match' => $revision) if $revision;
    $wipmeta->type($type);
    my $res=$wipmeta->put("/".$wipmeta->database."/$uid/$name",$data, {deserializer => 'application/json'});
    if ($res->code != 201) {
        if ($res->failed) {
            print STDERR "Content: " . $res->response->content;
        }
        die "put_attachment($uid) PUT of $name return code: ".$res->code."\n";
    }
    return $res->data->{rev};
}

1;


lib/CIHM/METS/App/Set.pm
package CIHM::METS::App::Set;

use common::sense;
use Data::Dumper;
use MooseX::App::Command;
use Try::Tiny;
use Cwd qw(realpath);
use File::Slurp;
use JSON;
use Text::CSV_XS;

extends qw(CIHM::METS::App);

parameter 'identifier' => (
  is => 'rw',
  isa => 'Str',
  required => 1,
  documentation => q[The identifier (Example: C-11 , debates_CDC2702_20)],
);

option 'setlabel' => (
  is => 'rw',
  isa => 'Str',
  documentation => q[String representing the label for this item],
);
option 'loadlabel' => (
  is => 'rw',
  isa => 'Str',
  documentation => q[A filename that will be loaded to set label for this item],
);

option 'componentlabels' => (
  is => 'rw',
  isa => 'Str',
  documentation => q[A filename read as component labels (.csv or .json)],
);
option 'dmdsec' => (
  is => 'rw',
  isa => 'Str',
  documentation => q[A filename containing the dmdSec for this item],
);
option 'metadata' => (
  is => 'rw',
  isa => 'Str',
  documentation => q[A filename containing a full METS record for this item],
);


command_short_description 'Sets metadata related fields and attachments in \'wipmeta\' database';


sub run {
  my ($self) = @_;

  $self->setup();

  # Load the provided label file
  if ($self->loadlabel) {
      my $label;
      open(my $fh,$self->loadlabel) or die "Can't open ".$self->loadlabel.": $!\n";
      defined($label = readline $fh) or die "Readline failed: $!";
      $label =~ s/\s+$//;
      if ($label ne '') {
          $self->setlabel($label);
      }
  }

  my $labels;
  if ($self->componentlabels) {
      my $file=$self->componentlabels;
      my ($ext) = $file =~ /(\.[^.]+)$/;
      if ($ext && $ext eq 'json') {
          my $clabels;
          try {
              $clabels = read_file($file);
          } catch {
              die "Error reading $file: $_";
          };
          try {
              $labels = from_json($clabels);
          } catch {
              die "Error parsing json file $file: $_\n";
          }
      } else {
          try {
              my @labels;
              my $csv = Text::CSV_XS->new ({ binary => 1});
              open my $fh, "<:encoding(utf8)", $file or die "Error opening $file: $!";
              while (my $row = $csv->getline ($fh)) {
                  push @labels,$row->[0];
              }
              close $fh;
              if (!($csv->eof)) {
                  my ($cde,$str,$pos, $rec, $fld) =  $csv->error_diag();
                  die "$str line $pos\n";
              }
              $labels = \@labels;
          } catch {
              die "Error parsing csv file $file: $_\n";
          }
      }
  }


  my $dmdsec;
  if ($self->dmdsec) {
      my $file=$self->dmdsec;
      try {
          $dmdsec = read_file($file);
      } catch {
          die "Error reading --dmdsec=$file: $_";
      };
  }

  my $metadata;
  if ($self->metadata) {
      my $file=$self->metadata;
      try {
          $metadata = read_file($file);
      } catch {
          die "Error reading --metadata=$file: $_";
      };
  }


  # Common function used to send them to CouchDB, no matter how they were
  # generated.
  $self->couchSend({
      label => $self->setlabel,
      clabels => $labels,
      dmdsec => $dmdsec,
      metadata => $metadata
                   });
  
}

1;


lib/CIHM/METS/App/Marc.pm
package CIHM::METS::App::Marc;

use common::sense;
use Encode;
use MooseX::App::Command;
use Try::Tiny;
use Cwd qw(realpath);
use File::Slurp;
use XML::LibXML;
use feature qw(say);
use Encode;
use MARC::Batch;
use MARC::File::XML (BinaryEncoding => 'utf8', RecordFormat => 'USMARC');
extends qw(CIHM::METS::App);

parameter 'marc' => (
  is => 'rw',
  isa => 'Str',
  required => 1,
  documentation => q[A .mrc file containing a metadata record],
);

option 'idschema' => (
  is => 'rw',
  isa => 'Str',
  default => '490',
  documentation => q[Identifier extraction schema (ooe|oocihm|490)],
);

option 'forceobjid' => (
  is => 'rw',
  isa => 'Str',
  documentation => q[Force this OBJID, and don't use idschema.  Only useful for single record marc files],
);


command_short_description 'Sets metadata related fields and attachments in \'wipmeta\' database';


sub run {
  my ($self) = @_;

  $self->setup();

  my $file = $self->marc ||
  die "Can't retrieve marc file\n";

  #process .mrc files
  my $marc = MARC::Batch->new("USMARC", $file);


  # Loop through each record
  while (my $record = $marc->next)  {

      my $label = $record->subfield(245, "a");

      if (!$label || $label =~/^\s*$/) {
          warn "Label can't be extracted from 245a\n";
          next;  # Don't try to store based on this record
      }

      if ($self->forceobjid) {
          $self->set_objid($self->forceobjid);
      }
      # Get the OBJID - based on legacy production/marcsplit
      elsif ($self->idschema eq 'ooe') {
          $self->set_objid(substr($record->subfield('035', 'a'),1));
      }
      elsif ($self->idschema eq '490' || $self->idschema eq 'oocihm') {
          my $objid;
          my $sf3=$record->subfield(490, "3");
          my $sfv=$record->subfield(490, "v");
          if ($sfv && $sfv ne '') {
              if ($sf3 && $sf3 ne '') {
                  $objid = join('_',$sf3,$sfv);
              } else {
                  $objid = $sfv;
                  if ($self->idschema eq 'oocihm') {
                      $objid =~ s/-/_/g;
                      $objid =~ s/[^0-9_]//g;
                  }
              }
              $self->set_objid($objid);
          } else {
              warn "490v is missing\n";
          }
      }
      else {
          die("Unknown identifier extraction scheme: ". $self->idschema . "\n");
      }

      if (!$self->objid) {
          warn "OBJID wasn't able to be determined\n";
          next;  # Don't try to store based on this record
      }

      say STDERR "processing: ".$self->objid;
      #create xml structure -- using the "encode_utf8" function to encode to Perl's internal format 
      my $xml = encode_utf8(join( "",
          MARC::File::XML::header(),
          MARC::File::XML::record($record),
          MARC::File::XML::footer()
      ));

      #send to couch
      $self->couchSend({
          label => $label,
          dmdsec =>   $xml
                       });
  }
}

1;


lib/CIHM/METS/App/csv2dc.pm
package CIHM::METS::App::csv2dc;

use common::sense;
use Data::Dumper;
use MooseX::App::Command;
use Try::Tiny;
use Cwd qw(realpath);
use File::Slurp;
use JSON;
use Text::CSV;
use XML::LibXML;
use List::Util qw(first);
use List::MoreUtils qw(first_index);
use File::Basename;
use feature qw(say);

extends qw(CIHM::METS::App);

parameter 'csv' => (
  is => 'rw',
  isa => 'Str',
  required => 1,
  documentation => q[A csv file containing metadata records],
);

command_short_description 'Sets metadata related fields and attachments in \'wipmeta\' database';

sub run {
    my ($self) = @_;

    $self->setup();

    my $file = $self->csv ||
        die "Can't retrieve csv file\n";

    # Where to find my XSD
    my $simpledcxsd = "/opt/xml/current/unpublished/xsd/simpledc.xsd";

    #process csv file
    my $csv = Text::CSV->new({binary=>1});
    open my $fh, "<:encoding(utf8)", $file or die "Cannot read $file: $!\n";

    my $header = $csv->getline($fh);
    #get object id
    my $objid_column = first {@$header[$_] && @$header[$_] eq 'objid'}0..@$header;
    if (!defined $objid_column) {
        die "column 'objid' header not found in first row\n"; 
    }

    while(my $row = $csv->getline($fh)) {
        #process each metadata record based on the object ID	
        foreach my $id ($row->[$objid_column]){
            if (!$id || $id =~ /^\s*$/) {
                warn "Line missing ID\n";
                next;
            }
		
            $self->set_objid($id);

            say STDERR "processing: ".$self->objid;
            my $doc = XML::LibXML::Document->new("1.0", "UTF-8");
            my $root = $doc->createElement('simpledc');
#            $root->setAttribute('xmlns:dc' => 'http://purl.org/dc/elements/1.1/');
            $root->setNamespace('http://purl.org/dc/elements/1.1/','dc',0);
            

            my %skipheaders = (
                'objid' => 1
            );

            # map header to dc element and process values	
            foreach my $thisheader (@$header) {
                if (substr($thisheader,0,3) eq 'dc:') {
                    get_element($thisheader, $header, $row, $root);
                }
            }
	  	
            #create xml file
            $doc->setDocumentElement($root);

            # TODO: This should be a NOP, but is required -- why?
            $doc= XML::LibXML->load_xml(string => $doc->toString(0));

            my $schema = XML::LibXML::Schema->new(location => $simpledcxsd);

            # Will die if validation failed.
            $schema->validate($doc);
            
            my $updatedoc = {
                dmdsec => $doc->toString(1)
            };

            my $label_col;
            my $mets_label;
            $label_col = first {@$header[$_] eq 'dc:title'}0..@$header;
            if ($label_col) {
                my $label_value = $row->[$label_col];

                #split on delimiters
                my @titles = split (/\s*\|\|\s*/, $label_value);
                $mets_label = $titles[0];
            }

            if ($mets_label) {
                $updatedoc->{label}= $mets_label,
            } else {
                warn "No METS label set!\n";
            }
            
            #send to couch
            $self->couchSend($updatedoc);
        }
    }
}

sub get_element{
	my ($header, $header_array, $row, $root) = @_;

        my $header_index = first {@$header_array[$_] eq $header}0..@$header_array;
        my $value = $row->[$header_index];
		
        #split on delimiters
        my @values = split (/\s*\|\|\s*/, $value);
        foreach (@values){
            $root->appendTextChild($header,$_);
        }	
}

1;


lib/CIHM/METS/App/csv2issue.pm
package CIHM::METS::App::csv2issue;

use common::sense;
use Data::Dumper;
use MooseX::App::Command;
use Try::Tiny;
use Cwd qw(realpath);
use File::Slurp;
use JSON;
use Text::CSV;
use XML::LibXML;
use List::Util qw(first);
use List::MoreUtils qw(first_index);
use File::Basename;
use feature qw(say);

extends qw(CIHM::METS::App);

parameter 'csv' => (
  is => 'rw',
  isa => 'Str',
  required => 1,
  documentation => q[A csv file containing metadata records],
);


command_short_description 'Sets metadata related fields and attachments in \'wipmeta\' database';


sub run {
    my ($self) = @_;

    $self->setup();

    my $label=$self->itemlabel;
    if (!$label) {
        warn "'itemlabel' not set for configid=".$self->configid." , will use default value 'column'\n";
        $label='column';
    }

    my $file = $self->csv ||
        die "Can't retrieve csv file\n";

    # Where to find my XSD
    my $issueinfoxsd = "/opt/xml/current/published/xsd/issueinfo.xsd";


    my @sequence = ('series','title','sequence','language','coverage','published','pubstatement','source','identifier','note');

    #process csv file
    my $csv = Text::CSV->new({binary=>1});
    open my $fh, "<:encoding(utf8)", $file or die "Cannot read $file: $!";
    my $headerline = $csv->getline($fh);

    # Create hash from headerline
    my %headers;
    my @unknownheader;
    for (my $i=0; $i < @$headerline; $i++) {
        my $header=$headerline->[$i];
        my $value={
            index => $i
        };
        if (index($header,'=') != -1) {
            my $type;
            ($header,$type) = split('=',$header);
            $value->{type}=$type;
        }

        if ($header eq 'objid' || $header eq 'label' ||
            first {$_ eq $header} @sequence) {
            if (!exists $headers{$header}) {
                $headers{$header}=[];
            }
            push $headers{$header}, $value;
        } else {
            push @unknownheader, $header;
        }
    }
    if (@unknownheader) {
        warn "The following headers are unknown: " . join(',',@unknownheader)."\n";
    }
    die "'title' header missing\n" if (!defined $headers{'title'});
    die "'label' header missing\n" if (!defined $headers{'label'});


    my %series;
    while(my $row = $csv->getline($fh)) {
	
        #get object id
        my $objid_column = first {@$headerline[$_] eq 'objid'}0..@$headerline;
	  
        #process each metadata record based on the object ID	
        foreach my $id ($row->[$objid_column]){
            if (!$id || $id =~ /^\s*$/) {
                warn "Line missing ID\n";
                next;
            }

            $self->set_objid($id);
            say STDERR "processing: ".$self->objid;

            my $doc = XML::LibXML::Document->new("1.0", "UTF-8");
            my $root = $doc->createElement('issueinfo');
            $root->setAttribute('xmlns' => 'http://canadiana.ca/schema/2012/xsd/issueinfo');	


            foreach my $element (@sequence) {
                if (exists $headers{$element}) {
                    foreach my $elementtype (@{$headers{$element}}) {
                        my $value = $row->[$elementtype->{'index'}];
                        if ($element eq "coverage") {
                            my $child= XML::LibXML::Element->new($element);
                            my $attribute=0;
                            if ($value =~ /start=([0-9-]+)/i) {
                                $child->setAttribute('start',$1);
                                $attribute++;
                            }
                            if ($value =~ /end=([0-9-]+)/i) {
                                $child->setAttribute('end',$1);
                                $attribute++;
                            }
                            $root->appendChild($child);

                            if (!$attribute) {
                                warn "Unable to parse coverage value: $value\n";
                            }
                        }
                        else {
                            my $type = $elementtype->{'type'};
                            
                            #split on delimiters
                            my @values = split (/\s*\|\|\s*/, $value);
                            foreach (@values) {
                                my $child= XML::LibXML::Element->new($element);
                                $child->appendTextNode($_);
                                if ($type) {
                                    $child->setAttribute('type',$type);
                                }
                                $root->appendChild($child);
                                if ($element eq 'series') {
                                    $series{$self->depositor.'.'.$_}=1;
                                }
                            }
                        }
                    }
                }
            }

            #create xml file
            $doc->setDocumentElement($root);

            # TODO: This should be a NOP, but is required -- why?
            $doc= XML::LibXML->load_xml(string => $doc->toString(0));

            my $schema = XML::LibXML::Schema->new(location => $issueinfoxsd);

            # Will die() if failed.
            $schema->validate($doc);
            		
            # Determine label
            my $label_col;
            my $mets_label;
            if ($label eq "column") {
                $mets_label = $row->[$headers{'label'}[0]->{'index'}];
            }
            else {
                die "not recognized label value: $label";
            }

            #send to couch
            $self->couchSend({
                label => $mets_label,
                dmdsec => $doc->toString (1),
                             });
        }
    }
    my @series=keys %series;
    my $serieserror=0;
    if ($self->WIP && @series) {
        my $wipmeta=$self->WIP->wipmeta;
        $wipmeta->type("application/json");
        my $res = $wipmeta->post("/".$wipmeta->{database}."/_all_docs?include_docs=true",{keys => \@series}, {deserializer => 'application/json'});
        if ($res->code == 200) {
            if (defined $res->data->{rows}) {
                foreach my $row (@{$res->data->{rows}}) {
                    if (exists $row->{error}) {
                        if ($serieserror++ == 0) {
                            print STDERR "\n\nSeries Warnings:\n";
                        }
                        warn "Series=".$row->{key}." : ".$row->{error}."\n";
                    } elsif (!exists $row->{doc}->{reposManifestDate}) {
                        if ($serieserror++ == 0) {
                            print STDERR "\n\nSeries Warnings:\n";
                        }
                        warn "Series=".$row->{key}." : exists, but no reposManifestDate field (not yet ingested)\n";
                    }
                }
            }
        }
        else {
            die "\nGET of series records returned: ".$res->response->status_line."\n"; 
        }
    }
}

1;


lib/CIHM/METS/App/dbtext2lac.pm
package CIHM::METS::App::dbtext2lac;

use common::sense;
use Data::Dumper;
use MooseX::App::Command;
use XML::LibXML;
use File::Basename;
use feature qw(say);

extends qw(CIHM::METS::App);

parameter 'dmp' => (
  is => 'rw',
  isa => 'Str',
  required => 1,
  documentation => q[An Inmagic DB/TextWorks file containing metadata records for LAC (reels and mikan)],
);


command_short_description 'Sets fields \'wipmeta\' database based on DB/Text xml dumps';


sub run {
    my ($self) = @_;

    $self->setup();

    my $file = $self->dmp ||
        die "Can't retrieve csv file\n";

    # Where to find my XSD
    $self->{simpledcxsd} = "/opt/xml/current/unpublished/xsd/simpledc.xsd";
    $self->{issueinfoxsd} = "/opt/xml/current/published/xsd/issueinfo.xsd";

    my $doc = XML::LibXML->load_xml(location => $file);
    $self->{xpc} = XML::LibXML::XPathContext->new($doc);
    $self->{xpc}->registerNs('inm', 'http://www.inmagic.com/webpublisher/query');

    # Used to store all parents, to be checked against wipmeta
    $self->{series}={};

    my @records=$self->xpc->findnodes('inm:Results/inm:Recordset/inm:Record');

    print "Processing ".scalar(@records)." records: \n\n"; 
    foreach my $record (@records) {
        $self->process_record($record);
    }

    my @series=keys $self->series;
    my $serieserror=0;
    if ($self->WIP && @series) {
        my $wipmeta=$self->WIP->wipmeta;
        $wipmeta->type("application/json");
        my $res = $wipmeta->post("/".$wipmeta->{database}."/_all_docs?include_docs=true",{keys => \@series}, {deserializer => 'application/json'});
        if ($res->code == 200) {
            if (defined $res->data->{rows}) {
                foreach my $row (@{$res->data->{rows}}) {
                    if (exists $row->{error}) {
                        if ($serieserror++ == 0) {
                            print STDERR "\n\nSeries Warnings:\n";
                        }
                        warn "Series=".$row->{key}." (".join(",",@{$self->series->{$row->{key}}}).") : ".$row->{error}."\n";
                    } elsif (!exists $row->{doc}->{reposManifestDate}) {
                        if ($serieserror++ == 0) {
                            print STDERR "\n\nSeries Warnings:\n";
                        }
                        warn "Series=".$row->{key}." (".join(",",@{$self->series->{$row->{key}}}).") : exists, but no reposManifestDate field (not yet ingested)\n";
                    }
                }
            }
        }
        else {
            die "\nGET of series records returned: ".$res->response->status_line."\n"; 
        }
    }
}

sub xpc {
    my $self = shift;
    return $self->{xpc};
}
sub series {
    my $self = shift;
    return $self->{series};
}


sub process_record {
    my($self,$record) = @_;

    my $doc = XML::LibXML::Document->new('1.0', 'UTF-8');
    my $type = $self->xpc->findvalue('inm:Type', $record);

    my $objid;
    my $label;

    if ($type eq 'Issue') {
        my $root = $doc->createElement('issueinfo');
        my $seq = $self->xpc->findvalue('inm:Reel_Number', $record);
        $objid = ($seq);
        $label = ($seq);
        $seq =~ s/[^0-9]//g;
        $objid =~ s/\W//g;
        
        #create lac reel objids
        $objid =~ s/(^[Cc])(.*)$/lac_reel_c$2/g;
        $objid =~ s/(^[Tt])(.*)$/lac_reel_t$2/g;
        $objid =~ s/(^[Hh])(.*)$/lac_reel_h$2/g;        

        say STDERR "$objid";
        
        $root->setAttribute('xmlns', 'http://canadiana.ca/schema/2012/xsd/issueinfo');
        $doc->setDocumentElement($root);

        my $series=join("_", "lac", "mikan", $self->xpc->findvalue('inm:Mikan', $record));
        my $seriesid=$self->depositor.".".$series;
        if (!exists $self->series->{$seriesid}) {
            $self->series->{$seriesid}=[];
        };
        push  $self->series->{$seriesid},$label;
        add_value($doc, 'series', $series);

        add_value($doc, 'title', join(' - ', $self->xpc->findvalue('inm:Title', $record), $seq));
        add_value($doc, 'sequence', $seq);
        add_field($doc, 'language', $self->xpc->findnodes('inm:Lang', $record));
        my $coverage_from = $self->xpc->findvalue('inm:Begin_Date', $record);
        my $coverage_to = $self->xpc->findvalue('inm:End_Date', $record);
        if ($coverage_from && $coverage_to) {
            my $coverage = $doc->createElement('coverage');
            $coverage->setAttribute('start', $coverage_from);
            $coverage->setAttribute('end', $coverage_to);
            $root->appendChild($coverage);
        }
        add_value($doc, 'source', "Library and Archives Canada / Bibliothèque et Archives Canada");
        foreach my $identifier ($self->xpc->findnodes('inm:Reel_Number', $record)) {
            my $value = $identifier->findvalue('.');
            next unless($value);
            my $node = $doc->createElement('identifier');
            $node->setAttribute('type', 'mikan_reel');
            $node->appendChild($doc->createTextNode($value));
            $root->appendChild($node);
        }
        foreach my $identifier ($self->xpc->findnodes('inm:Mikan', $record), $self->xpc->findnodes('inm:Mikan_F', $record), $self->xpc->findnodes('inm:Mikan_Other', $record)) {
            my $value = $identifier->findvalue('.');
            next unless($value);
            my $node = $doc->createElement('identifier');
            $node->setAttribute('type', 'mikan_number');
            $node->appendChild($doc->createTextNode($value));
            $root->appendChild($node);
        }
        foreach my $identifier ($self->xpc->findnodes('inm:Reference', $record)) {
            my $value = $identifier->findvalue('.');
            next unless($value);
            my $node = $doc->createElement('identifier');
            $node->setAttribute('type', 'mikan_reference');
            $node->appendChild($doc->createTextNode($value));
            $root->appendChild($node);
        }


        # TODO: This should be a NOP, but is required -- why?
        $doc= XML::LibXML->load_xml(string => $doc->toString(0));

        my $schema = XML::LibXML::Schema->new(location => $self->{issueinfoxsd});

        # Will die() if failed.
        $schema->validate($doc);

        
    }
    elsif ($type eq 'Series') {
        my $root = $doc->createElement('simpledc');
        my $mikan = $self->xpc->findvalue('inm:Mikan', $record);

        $objid = "lac_mikan_$mikan";
        say STDERR "$objid";

        # Label is the first title
        $label = $self->xpc->findnodes('inm:Title', $record)->[0]->findvalue('.');
        $root->setAttribute('xmlns:dc', 'http://purl.org/dc/elements/1.1/');
        $doc->setDocumentElement($root);
        add_value($doc, 'dc:identifier', join("_", "lac", "mikan", $mikan));
        add_field($doc, 'dc:title', $self->xpc->findnodes('inm:Title', $record));
        add_field($doc, 'dc:title', $self->xpc->findnodes('inm:AT', $record));
        add_field($doc, 'dc:date', $self->xpc->findnodes('inm:Begin_Date', $record));
        add_field($doc, 'dc:date', $self->xpc->findnodes('inm:End_Date', $record));
        add_field($doc, 'dc:language', $self->xpc->findnodes('inm:Lang', $record));
        add_field($doc, 'dc:identifier', $self->xpc->findnodes('inm:Mikan_F', $record));
        add_field($doc, 'dc:identifier', $self->xpc->findnodes('inm:Mikan_Other', $record));
        add_field($doc, 'dc:identifier', $self->xpc->findnodes('inm:Reference', $record));
        add_field($doc, 'dc:identifier', $self->xpc->findnodes('inm:Archival_Reference', $record));
        add_field($doc, 'dc:identifier', $self->xpc->findnodes('inm:Collections', $record));
        add_field($doc, 'dc:subject', $self->xpc->findnodes('inm:Subject', $record));
        add_field($doc, 'dc:creator', $self->xpc->findnodes('inm:AU', $record));
        add_field($doc, 'dc:creator', $self->xpc->findnodes('inm:AN', $record));
        add_field($doc, 'dc:description', $self->xpc->findnodes('inm:Description', $record));
        add_field($doc, 'dc:description', $self->xpc->findnodes('inm:Contents', $record));    
        add_field($doc, 'dc:description', $self->xpc->findnodes('inm:Contents_F', $record));
        add_value($doc, 'dc:source', "Library and Archives Canada / Bibliothèque et Archives Canada");


        # TODO: This should be a NOP, but is required -- why?
        $doc= XML::LibXML->load_xml(string => $doc->toString(0));

        my $schema = XML::LibXML::Schema->new(location => $self->{simpledcxsd});

        # Will die() if failed.
        $schema->validate($doc);
        
    }
    else {
        die("Cannot process record with unknown type \"$type\"\n");
    }

    $self->set_objid($objid);

    my $updatedoc = {
        label => $label,
        dmdsec => $doc->toString(1)
    };

    #send to couch
    $self->couchSend($updatedoc);
}

sub add_value {
    my($doc, $field, $value) = @_;
    return unless ($value);
    my $node = $doc->createElement($field);
    $node->appendChild($doc->createTextNode($value));
    $doc->documentElement->appendChild($node);
}

sub add_field {
    my($doc, $field, @list) = @_;
    foreach my $element (@list) {
        my $value = $element->findvalue('.');
        next unless ($value);
        my $text = $doc->createTextNode($value);
        my $node = $doc->createElement($field);
        $node->appendChild($text);
        $doc->documentElement->appendChild($node);
    }
}




1;
